### PR TITLE
fix autoescape for non-string values; fixes #835

### LINKED ISSUE #835: Bypass autoescape / XSS
The following string works as expected: 

``` javascript
var res = nunjucks.renderString('Hello {{ username }}', { username: '<s>Matt</s>' });
console.log(res); //Hello &lt;s&gt;Matt&lt;/s&gt;
```

If however the variable passed to the template is an array autoescape does nothing: 

``` javascript
var res = nunjucks.renderString('Hello {{ username }}', { username: ['<s>Matt</s>'] });
console.log(res); // Hello <s>Matt</s>
```

This looks to be intentional https://github.com/mozilla/nunjucks/blob/master/src/runtime.js#L209 
However if a autoescape is on any variable that is rendered in a {{ }} block and appended to the output I would expect it to escaped no matter the type.  If the var is just going to be part of a string concat anyway why not toString first then escape?  

In express / Koa / (anything else using qs or body-parser) is is trivial to coerce query params types. See the following simple example in express: 

``` javascript
var app = express();
var env = nunjucks.configure('views', {
    autoescape: true,
    express: app
});

app.get('/', function(req, res) {     
    // This should be fine autoescape is on...
    res.send(nunjucks.renderString('Hello {{ username }}', { username: req.query.name }));
    //res.render('index.html', { username: req.query.name }); 
});
```
`http://127.0.0.1:3000/?name[]=<script>alert(1)</script>matt`

I created a more detailed writeup and example app at:  https://github.com/matt-/nunjucks_test

### PR DESCRIPTION
**Pull request by devoidfury**
#835

### CODE REVIEW

**Comment by devoidfury** [comment]
Imo, a bugfix for the issue is release asap material (even if it's not this one). Maybe submit a CVE?

@SamyPesse @jlongster @vecmezoni  @carljm @jbmoelker

File: src/runtime.js [hunk 1]
```diff
@@ -206,8 +206,8 @@ function markSafe(val) {
 function suppressValue(val, autoescape) {
     val = (val !== undefined && val !== null) ? val : '';
 
-    if(autoescape && typeof val === 'string') {
-        val = lib.escape(val);
+    if(autoescape && !(val instanceof SafeString)) {
+        val = lib.escape(''+val);
```
Comment (vecmezoni) [inline src/runtime.js]: `val.toString()` will be clearer

**Comment by jbmoelker** [comment]
@devoidfury thanks for this! Looks solid. Could you update the `CHANGELOG.md` with the fix? For me that's all that's left before merging this PR.

I don't know if we need to submit a CVE. I understand the risk of this bug. But I don't know the protocol or procedure.

@carljm what's the procedure to release this? I've read `MAINTENANCE.md`, but I don't quite understand how the `2.x` and `3.x` branches are managed. Seeing the [unreleased 2.x items](https://github.com/mozilla/nunjucks/blob/master/CHANGELOG.md#2x-unreleased) I imagine we should release `2.5.0`. Also I think @vecmezoni and I as 1 day old maintainers are able to create releases on GitHub, but are unable to publish a new version to npm.

**Comment by vecmezoni** [comment]
Probably would be better to publish 2.4.3 and 2.5.0 to fix vulnerability, so users with all the cases of package.json versioning `~ ^` will receive fix. Just need to check is there anything to bump a minor version or we might bump just patch (2.4.3)

**Comment by jbmoelker** [comment]
@vecmezoni well there are multiple [unreleased features on 2.x](https://github.com/mozilla/nunjucks/blob/master/CHANGELOG.md#2x-unreleased), so that's why the minor bump. I agree that normally we'd merge this PR and release patch. But I'm not sure where to merge onto for that patch release.

**Comment by vecmezoni** [comment]
We can make separate branch from 2.4.2 version, apply this patch and release 2.4.3. Then we can apply this patch to 2.x and decide to release or not to release 2.5. I've asked @carljm to give me access to npm, waiting.

**Comment by vecmezoni** [comment]
In this case people with `"nunjucks": "^2.4"` will get this update. Also we can add deprecation warning for everything less than 2.4.2 https://docs.npmjs.com/cli/deprecate

**Comment by vecmezoni** [comment]
@devoidfury add note to `CHANGELOG.md` and i'm going to merge this and prepare for release and deprecation note.

**Comment by matt-** [comment]
Should the escape filter also do the same?

**Comment by devoidfury** [comment]
Yes, thanks @matt- -- added another commit to fix this.

File: CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +26,9 @@ Changelog
 * Fix handling of macro arg with default value which shares a name with another
   macro. Merge of [#791](https://github.com/mozilla/nunjucks/pull/791).
 
+* Fix potential cast-related XSS vulnerability in autoescape mode, and with `escape` filter.
```
Comment (jbmoelker) [inline CHANGELOG.md]: Can be merged without, but credits where due: "Thanks Matt Austin and Thomas Hunkapiller.".

**Comment by vecmezoni** [comment]
👍  Merging this to master, then i'll create branch 2.4.3 from 2.4.2 release, backport this patch and rebase 2.x branch on 2.4.3.

**Comment by carljm** [comment]
@vecmezoni sounds good, but I'd avoid rebasing the `2.x` branch. It's public, that could really mess up anyone else tracking it. just merge from your new 2.4.3 branch into 2.x instead.

**Comment by vecmezoni** [comment]
* Released 2.4.3
* Added deprecation note (i'm not sure that vulnerability is critical, but anyways)
```
> npm i nunjucks@2.4.2

npm WARN deprecated nunjucks@2.4.2: potential XSS vulnerability in autoescape mode, 
and with escape filter was fixed in v2.4.3
```
* Backported patch to 2.x, tomorrow i will review and test branch and issue 2.5.0 release.

**Comment by dlmr** [comment]
Seems like this change broke the use of the `safe` filter. 

```
{{ value | safe }}
```

With `2.4.3` this will be escaped where I thought it should not be that. Could be that I have used it incorrectly in my application and that this was never supposed to work but wanted to give a heads up if this was not the case.

**Comment by devoidfury** [comment]
@dlmr Could you provide a simple test case here?

**Comment by dlmr** [comment]
@devoidfury I think I found the problem when trying to create a minimal example.

I was using [React Helmet](https://github.com/nfl/react-helmet) but did not use the `toString()` function on what I got back from it which means that the safe filter did nothing in that case and then when rendering the escape did not work, since it was not a string, and I got the unescaped output as I wanted.

So I was actually this bug in my code. The more correct thing is of course to use `toString` and in the process be more correct. 

Thanks for the quick reply and the fix in the first place! Sorry to blame the PR for this! 😄

**Comment by nelsonpecora** [comment]
Haha thanks @dimr for catching this. I just hit this with [an old filter we were using](https://github.com/nymag/nunjucks-filters/blob/master/filters/files/includeFile.js) for embedding optional svgs. It definitely woke me up this morning!

screen shot 2016-09-08 at 11 58 41 am

If I'm parsing the tests in this PR correctly, I should be doing something like this in my filter, correct?

```js
'use strict';
var fs = require('fs');

module.exports = function (path) {
  try {
    return {
      toString: function () {
        return fs.readFileSync(path);
      }
    };
  } catch () {
    return '';
  }
};
```

**Comment by devoidfury** [comment]
@vecmezoni @carljm should we modify the SafeString constructor to automatically do a string conversion? Right now, it's no-op (identity) on non-strings, but after that point it's going to be cast to string anyway, and with autoescape, if it's not a SafeString, it will escape.

**Comment by fengmk2** [comment]
Should we fix this in 1.x?

File: src/filters.js [hunk 3]
```diff
@@ -116,10 +116,10 @@ var filters = {
     },
 
     escape: function(str) {
-        if(typeof str === 'string') {
-            return r.markSafe(lib.escape(str));
+        if(str instanceof r.SafeString) {
+            return str;
         }
-        return str;
+        return r.markSafe(lib.escape(str.toString()));
```
Comment (fengmk2) [inline src/filters.js]: this will throw `TypeError: Cannot read property 'toString' of null` when str is `null` or `undefined`

File: tests/compiler.js [hunk 4]
```diff
@@ -1395,6 +1395,14 @@
                 expect(res).to.be('&><\'"');
             });
 
+            render('{{ foo }}', { foo: ['<p>foo</p>']}, { autoescape: true }, function(err, res) {
```
Comment (fengmk2) [inline tests/compiler.js]: Add one more test case for `null` value check:

```js
render('{{ foo }}', { foo: null }, { autoescape: true }, function(err, res) {
  // ...
});
```

File: tests/filters.js [hunk 5]
```diff
@@ -110,6 +110,15 @@
             finish(done);
         });
 
+        it('should work with non-string values', function(done) {
+            var res1 = render('{{ foo | escape }}', {foo: ['<html>']}, { autoescape: false });
```
Comment (fengmk2) [inline tests/filters.js]: Add one more test case for null value check.

File: src/filters.js (hunk 3 above)
Comment (atian25) [inline src/filters.js]: https://github.com/mozilla/nunjucks/pull/843
